### ISSUE TITLE
[CWM]  Add the IFFCO membership data to the CWM.

### ISSUE THREAD

**Issue by lin-d-hop** [DigitalCommons/mykomap-monolith#47]
## Description

Add the IFFCO membership data to the CWM.
[This spreadsheet](https://docs.google.com/spreadsheets/d/1bIVGZ1O56NVURZ9Zw7yDdz2sALTKEWgVxhYwY3hwwfM/edit?gid=0#gid=0) lists all the datasets and their source files

This will involve:
- normalising the data
- merging the data

## Acceptance Criteria

- IFFCO membership data is live on the CWM

**Comment by ColmDC** [comment]
https://github.com/DigitalCommons/cwm-test-data/issues/1

**Comment by wu-lee** [comment]
This a sub-issue of #54 I think...

**Comment by wu-lee** [comment]
Mapbox geocoded data now deployed here https://dev.data.solidarityeconomy.coop/iffco/

John recommended mapbox as being the better choice than geoapify for this dataset, at least. 

It's pretty big, so takes hours to geocode!

**Comment by ColmDC** [comment]
Now that we have the data in the CWM we can see how many are way off!

https://dev.maps.coop/cwm/?datasetId=delhi

Select India in the Directory and it's easy to see about 70 spread across the rest of the world.

@John Can we check what confidence Mapbox gave for these ones? Is there a threshold where we can treat them as not having geocodes?

**Comment by King-Mob** [comment]
@ColmDC yes I am

**Comment by ColmDC** [comment]
Is IFFCO now using the improved geocoded data?

**Comment by wu-lee** [comment]
The datafactory gets something it interprets as a confidence from mapbox when it's using that for geocoding.  Let me check...   it uses a value called "relevance"

The docs [here][1] just say the results are "ordered by relevance" but don't mention it as an attribute.  The geocoding is performed using the [`geocoder`][2] Gem, but the docs and source code for that don't say much about the "relevance" parameter except that I can see that it sorts results by it.


[1]: https://docs.mapbox.com/api/search/geocoding/
[2]: https://github.com/alexreisner/geocoder

**Comment by wu-lee** [comment]
Ok, found a page on Mapbox's site describing `relevance`: 

https://docs.mapbox.com/help/getting-started/geocoding/

This suggests that relevance is a useful equivalent for confidence. We don't include the `score` value mentioned, but that seems to mainly be relevant for things which are commonly searched for, which may not add much to an Indian Agricultural Co-op address, as I think those will be rarely searched for.  And our own geocoding might skew the results...

> ### Result prioritization in forward geocoding
> 
> When a forward geocoding query (human-readable text like “San Francisco”) is submitted to the Mapbox Geocoding API, the geocoder applies filters on the backend to sort the results that match, or partially match, the searched text. These filters are textual relevance and prominence score.
> Textual relevance
> 
> When the geocoder sorts and prioritizes results, the first filter that is applied is relevance. relevance is a value that indicates how well a feature in our dataset matches the query. This property is surfaced in the Geocoding API response, and is a numerical score from 0 (the result does not match the queried text at all) to 1 (the result matches the queried text most completely). You can use the relevance property to remove results that don't fully match the query.
> 
> In this example query for the address “515 15th St NW, Washington, DC 20004”, the expected result is first in the response with a relevance of 0.875. Other results in other cities, since they don’t match the search text as closely, have a relevance score of 0.2.
> 
>     https://api.mapbox.com/geocoding/v5/mapbox.places/515%2015th%20St%20NW%2C%20Washington%2C%20DC%2020004.json?types=address&access_token=YOUR_MAPBOX_ACCESS_TOKEN
> 
> ### Prominence score
> 
> In the case that multiple features have the same relevance score, a second filter called score is applied. This is based on the popularity or prominence of a feature. For example, a search for “Paris” will equally match “Paris, France” and “Paris, Texas” — they’ll have the same relevance score. The score filter helps break this tie on the backend, and surfaces “Paris, France” first since it is a more popular feature:
> 
>     https://api.mapbox.com/geocoding/v5/mapbox.places/paris.json?access_token=YOUR_MAPBOX_ACCESS_TOKEN
> 
>

**Comment by wu-lee** [comment]
> Is IFFCO now using the improved geocoded data?

Just to answer this, it's currently using the Mapbox data, but not attempting to do structured queries as described, because that would require ditching all geocoding code being used currently and inserting custom mapbox queries.

If the reason for this is to get confidence values, we already get them. Checking the data, of all the 35k IFFCO rows, 3.8k are NULL, 8.0k are 0.  (Table below)

If the reason is to improve the accuracy, maybe it's worth it in a longer scale - not sure if the current timescale allows it. If we do that, perhaps the relevance can be used to assess that.

|field|value|count| frequency bar |
|--|--|--|--|
|Geo Container Confidence|(NULL)|3770|-------------------------------------|
|Geo Container Confidence|0|7967|-------------------------------------------------------------------------------|
|Geo Container Confidence|1|206|--|
|Geo Container Confidence|2|113|-|
|Geo Container Confidence|3|96||
|Geo Container Confidence|4|68||
|Geo Container Confidence|5|96||
|Geo Container Confidence|6|61||
|Geo Container Confidence|7|63||
|Geo Container Confidence|8|52||
|Geo Container Confidence|9|33||
|Geo Container Confidence|10|77||
|Geo Container Confidence|11|60||
|Geo Container Confidence|12|448|----|
|Geo Container Confidence|13|37||
|Geo Container Confidence|14|78||
|Geo Container Confidence|15|31||
|Geo Container Confidence|16|81||
|Geo Container Confidence|17|75||
|Geo Container Confidence|18|154|-|
|Geo Container Confidence|19|93||
|Geo Container Confidence|20|79||
|Geo Container Confidence|21|61||
|Geo Container Confidence|22|116|-|
|Geo Container Confidence|23|37||
|Geo Container Confidence|24|27||
|Geo Container Confidence|25|12145|-------------------------------------------------------------------------------------------------------------------------|
|Geo Container Confidence|26|15||
|Geo Container Confidence|27|26||
|Geo Container Confidence|28|21||
|Geo Container Confidence|29|6||
|Geo Container Confidence|30|22||
|Geo Container Confidence|31|27||
|Geo Container Confidence|32|18||
|Geo Container Confidence|33|37||
|Geo Container Confidence|34|16||
|Geo Container Confidence|35|14||
|Geo Container Confidence|36|26||
|Geo Container Confidence|37|27||
|Geo Container Confidence|38|5||
|Geo Container Confidence|39|5||
|Geo Container Confidence|40|82||
|Geo Container Confidence|41|13||
|Geo Container Confidence|42|12||
|Geo Container Confidence|43|6||
|Geo Container Confidence|44|21||
|Geo Container Confidence|45|255|--|
|Geo Container Confidence|46|10||
|Geo Container Confidence|47|13||
|Geo Container Confidence|48|17||
|Geo Container Confidence|49|7||
|Geo Container Confidence|50|3123|-------------------------------|
|Geo Container Confidence|51|94||
|Geo Container Confidence|52|80||
|Geo Container Confidence|53|76||
|Geo Container Confidence|54|107|-|
|Geo Container Confidence|55|81||
|Geo Container Confidence|56|159|-|
|Geo Container Confidence|57|169|-|
|Geo Container Confidence|58|83||
|Geo Container Confidence|59|59||
|Geo Container Confidence|60|861|--------|
|Geo Container Confidence|61|106|-|
|Geo Container Confidence|62|116|-|
|Geo Container Confidence|63|208|--|
|Geo Container Confidence|64|100|-|
|Geo Container Confidence|65|112|-|
|Geo Container Confidence|66|312|---|
|Geo Container Confidence|67|96||
|Geo Container Confidence|68|96||
|Geo Container Confidence|69|84||
|Geo Container Confidence|70|126|-|
|Geo Container Confidence|71|90||
|Geo Container Confidence|72|109|-|
|Geo Container Confidence|73|71||
|Geo Container Confidence|74|61||
|Geo Container Confidence|75|90||
|Geo Container Confidence|76|59||
|Geo Container Confidence|77|55||
|Geo Container Confidence|78|45||
|Geo Container Confidence|79|38||
|Geo Container Confidence|80|401|----|
|Geo Container Confidence|81|163|-|
|Geo Container Confidence|82|34||
|Geo Container Confidence|83|36||
|Geo Container Confidence|84|17||
|Geo Container Confidence|85|26||
|Geo Container Confidence|86|20||
|Geo Container Confidence|87|10||
|Geo Container Confidence|88|21||
|Geo Container Confidence|89|7||
|Geo Container Confidence|90|381|---|
|Geo Container Confidence|91|10||
|Geo Container Confidence|92|15||
|Geo Container Confidence|93|10||
|Geo Container Confidence|94|29||
|Geo Container Confidence|95|52||
|Geo Container Confidence|96|12||
|Geo Container Confidence|97|1||
|Geo Container Confidence|100|687|------|

**Comment by wu-lee** [comment]
But then having said that, I've just ran the geoapify-data through the merge and convert process so I can look visually at the distribution if the IFFCO pins. And GeoAPIfy looks better as it doesn't put so many pins outside India.

This is with mapbox, i.e. the currently deployed data:

![image](https://github.com/user-attachments/assets/258ea375-254c-4da5-a0de-b956ef9cfb9b)

And this is with the GeoAPIfy geocoded IFFCO data:

![image](https://github.com/user-attachments/assets/d59deb36-e038-4489-b02f-ba0711a5e104)

**Comment by wu-lee** [comment]
I've switched the published version if the IFFCO dataset to the geoAPIfy geocoded version now. (I have two versions of this data stored on my desktop, one for Mapbox and one for GeoAPIfy)

https://dev.data.solidarityeconomy.coop/iffco/standard.csv

This means when I ran the demo-merge-map-data script `generate-db2.sh` it pulled this version, and the latest data deployment includes it. So when I check, and filter the pins to IFFCO members, I see the second image above.
